Create new task for building basic daemon set (#48655)
kubernetes/website: content/en/docs/tasks/manage-daemon/create-daemon-set.md

Reviewer: Make sure to:

- use an official pause image
- use a `code_sample` shortcode - see https://github.com/kubernetes/website/pull/48655/files#r1832177630
- include clean up steps
- write a _task_ rather than a tutorial

---
title: Building a Basic DaemonSet  
content_type: task  

> weight: 30  

Reviewer 2: ```suggestion
weight: 5
```
As this page explains about building a basic DaemonSet, it must be the first task.

---
<!-- overview -->

This page demonstrates how to build a basic DaemonSet that runs a Pod on every node in a Kubernetes cluster. It covers a simple use case of mounting a file from the host, logging its contents using an init container, and utilizing a pause container.

## {{% heading "prerequisites" %}}

{{< include "task-tutorial-prereqs.md" >}}

## Building a DaemonSet

In this task, you'll create a basic DaemonSet that mounts the `/etc/machine-id` file from each node, logs its contents using an init container, and runs a pause container.

### Step 1: Create the DaemonSet Manifest

> Let's create a DaemonSet manifest that ensures one Pod is scheduled on every node. The Pod will use an init container to read and log the contents of `/etc/machine-id` from the host, while the main container will be a `pause` container, which keeps the Pod running.

Reviewer 2: ```suggestion
Create a DaemonSet manifest that ensures the copy of a Pod is scheduled on every node.
The Pod will use an init container to read and log the contents of `/etc/machine-id` from the host,
while the main container will be a `pause` container, which keeps the Pod running.
```

```yaml
apiVersion: apps/v1
kind: DaemonSet
metadata:
  name: example-daemonset
spec:
  selector:
    matchLabels:
      app: example
  template:
    metadata:
      labels:
        app: example
    spec:
      containers:
      - name: pause
        image: gcr.io/google_containers/pause:3.1
        # This is the main container, which doesn't do much but keeps the Pod alive.
      initContainers:
      - name: log-machine-id
        image: busybox
        command: ['sh', '-c', 'cat /etc/machine-id > /var/log/machine-id.log']
        volumeMounts:
        - name: machine-id
          mountPath: /etc/machine-id
          readOnly: true
        - name: log-dir
          mountPath: /var/log
      volumes:
      - name: machine-id
        hostPath:
          path: /etc/machine-id
          type: File
      - name: log-dir
        hostPath:
          path: /var/log
```

### Step 2: Apply the DaemonSet

> Apply the DaemonSet to your Kubernetes cluster by saving the manifest to a file (For example, `daemonset.yaml`) and using the following command:

Reviewer 2: Instead of mentioning raw yaml script, its good if you can place the yaml file in code samples and reference here.

Author: @T-Lakshmi I wanted to understand so here we should explain the particular code blocks instead of giving the whole yaml ?

Author: In the task doc I see usually the raw yaml is mentioned for example https://kubernetes.io/docs/tasks/manage-daemon/update-daemon-set/ 
I tried to follow the same structure, can you please elaborate the new change if possible that would be really helpful ?

Reviewer: Have a look at https://kubernetes.io/search/?q=code_sample; that should help you find advice on what to do.

You can - and should - look at how other task pages tackle this.

```bash
kubectl apply -f daemonset.yaml
```

### Step 3: Verify the DaemonSet

Once applied, you can verify that the DaemonSet is running a Pod on every node in the cluster:

```bash
kubectl get pods -o wide
```

The output will list one Pod per node, similar to:

```
NAME                                READY   STATUS    RESTARTS   AGE    IP       NODE
example-daemonset-xxxxx             1/1     Running   0          5m     x.x.x.x  node-1
example-daemonset-yyyyy             1/1     Running   0          5m     x.x.x.x  node-2
```

### Step 4: Inspect the Log

You can inspect the contents of the logged `/etc/machine-id` file by checking the log directory mounted from the host:

```bash
kubectl exec <pod-name> -- cat /var/log/machine-id.log
```

> 

Reviewer 2: ```suggestion
Replace <pod-name> with actual name of the pod running.

```

### Key Features of DaemonSets

DaemonSets provide several features for managing Pods on a per-node basis:

> **Host Path Volume Mounts**: DaemonSets can mount host directories (like `/etc/machine-id`) into Pods to access system-level files or logs.

Reviewer 2: ```suggestion
**Host Path Volume Mounts**:
DaemonSets can mount host directories (like `/etc/machine-id`) into Pods to access system-level files or logs.
```

> **Init Containers**: Init containers can perform setup tasks before the main container starts, such as logging or preparing the environment.

Reviewer 2: ```suggestion
**Init Containers**:
Init containers can perform setup tasks before the main container starts, such as logging or preparing the environment.
```

> **Pause Containers**: You can use a pause container as the main container when the Pod's primary function is handled by an init container, but you still want the Pod to remain running.

Reviewer 2: ```suggestion
**Pause Containers**:
You can use a pause container as the main container when the Pod's primary function is
handled by an init container, but you still want the Pod to remain running.
```

> This simple DaemonSet example introduces key components like init containers and host path volumes, which can be expanded upon for more advanced use cases.

Reviewer 2: ```suggestion
This simple DaemonSet example introduces key components like init containers and host path volumes,
which can be expanded upon for more advanced use cases. For more details refer
[DaemonSet](/docs/concepts/workloads/controllers/daemonset/). 
```

> 

Reviewer 2: ```suggestion

```

> 

Reviewer 2: ```suggestion

```

> 

Reviewer 2: ```suggestion

```

---
title: Building a Basic DaemonSet  
content_type: task  
weight: 5  
---
<!-- overview -->

This page demonstrates how to build a basic DaemonSet that runs a Pod on every node in a Kubernetes cluster.
It covers a simple use case of mounting a file from the host, logging its contents using
an [init container](/docs/concepts/workloads/pods/init-containers/), and utilizing a pause container.

## {{% heading "prerequisites" %}}

{{< include "task-tutorial-prereqs.md" >}}

## Building a DaemonSet

In this task, you'll create a basic DaemonSet that mounts the `/etc/machine-id` file from each node,
logs its contents using an init container, and runs a pause container.

### Step 1: Create the DaemonSet Manifest

Create a DaemonSet manifest that ensures the copy of a Pod is scheduled on every node.
The Pod will use an init container to read and log the contents of `/etc/machine-id` from the host,
while the main container will be a `pause` container, which keeps the Pod running.

```yaml
apiVersion: apps/v1
kind: DaemonSet
metadata:
  name: example-daemonset
spec:

>   selector:
>     matchLabels:
>       app: example
>   template:
>     metadata:
>       labels:
>         app: example

Reviewer: ```suggestion
  selector:
    matchLabels:
      app.kubernetes.io/name: example
  template:
    metadata:
      labels:
        app.kubernetes.io/name: example
```

      initContainers:
      - name: log-machine-id
        image: busybox
        command: ['sh', '-c', 'cat /etc/machine-id > /var/log/machine-id.log']
        volumeMounts:
        - name: machine-id
          mountPath: /etc/machine-id
          readOnly: true
        - name: log-dir
          mountPath: /var/log
      volumes:
      - name: machine-id
        hostPath:
          path: /etc/machine-id
          type: File
      - name: log-dir
        hostPath:
          path: /var/log
```

### Step 2: Apply the DaemonSet

Apply the DaemonSet to your Kubernetes cluster by saving the manifest to a file (For example, `daemonset.yaml`) and using the following command:

```bash
kubectl apply -f daemonset.yaml
```

### Step 3: Verify the DaemonSet

Once applied, you can verify that the DaemonSet is running a Pod on every node in the cluster:

```bash
kubectl get pods -o wide
```

The output will list one Pod per node, similar to:

```
NAME                                READY   STATUS    RESTARTS   AGE    IP       NODE
example-daemonset-xxxxx             1/1     Running   0          5m     x.x.x.x  node-1

> example-daemonset-yyyyy             1/1     Running   0          5m     x.x.x.x  node-2

Reviewer: How many nodes does this cluster have? Should we mention something in prerequisites?

```

### Step 4: Inspect the Log

You can inspect the contents of the logged `/etc/machine-id` file by checking the log directory mounted from the host:

```bash
kubectl exec <pod-name> -- cat /var/log/machine-id.log
```
Replace <pod-name> with actual name of the pod running.

> ### Key Features of DaemonSets

Reviewer: Either omit this, or move it to the start of the task.

I recommend linking to the DaemonSet concept page.

Author: I've removed this part of key features

---
title: Building a Basic DaemonSet  
content_type: task  
weight: 5  
---
<!-- overview -->

> This page demonstrates how to build a basic DaemonSet that runs a Pod on every node in a Kubernetes cluster.

Reviewer: ```suggestion
This page demonstrates how to build a basic {{< glossary_tooltip text="DaemonSet" term_id="daemonset" >}} that runs a Pod on every node in a Kubernetes cluster.
```

It covers a simple use case of mounting a file from the host, logging its contents using
an [init container](/docs/concepts/workloads/pods/init-containers/), and utilizing a pause container.

## {{% heading "prerequisites" %}}

{{< include "task-tutorial-prereqs.md" >}}

> A Kubernetes cluster with at least two nodes (one master node and one worker node) to demonstrate the behavior of DaemonSets.

Reviewer: ```suggestion
A Kubernetes cluster with at least two nodes (one control plane node and one worker node) to demonstrate the behavior of DaemonSets.
```

Author: I've addressed the comments @sftim!

> ## Building a DaemonSet

Reviewer: ```suggestion
## Define the DaemonSet
```

In this task, a basic DaemonSet is created which ensures that the copy of a Pod is scheduled on every node.
The Pod will use an init container to read and log the contents of `/etc/machine-id` from the host,
while the main container will be a `pause` container, which keeps the Pod running.

{{% code_sample file="application/basic-daemonset.yaml" %}}

> 1. Create a Daemonset based on the YAML file:

Reviewer: (nit)
```suggestion
1. Create a DaemonSet based on the (YAML) manifest:
```

  ```shell
  kubectl apply -f https://k8s.io/examples/application/basic-daemonset.yaml
  ```

1. Once applied, you can verify that the DaemonSet is running a Pod on every node in the cluster:

  ```shell
  kubectl get pods -o wide
  ```

  The output will list one Pod per node, similar to:

  ```
  NAME                                READY   STATUS    RESTARTS   AGE    IP       NODE
  example-daemonset-xxxxx             1/1     Running   0          5m     x.x.x.x  node-1
  example-daemonset-yyyyy             1/1     Running   0          5m     x.x.x.x  node-2
  ```

1. You can inspect the contents of the logged `/etc/machine-id` file by checking the log directory mounted from the host:

  ```shell
  kubectl exec <pod-name> -- cat /var/log/machine-id.log
  ```
  

>   where `<pod-name>` is the name of one of your Pod.

Reviewer 3: ```suggestion
  Where `<pod-name>` is the name of one of your Pods.
```

Author: updated the change!

Reviewer 3: Nice!

> 1. To clean up the Daemonset and its Pods:

Reviewer: ```suggestion
## {{% heading "cleanup" %}}

To clean up the DaemonSet (and [garbage collect](/docs/concepts/architecture/garbage-collection/) its Pods automatically):
````

  ```

>   kubectl delete -f basic-daemonset.yaml

Reviewer: This doesn't look right. Try this instead:
```suggestion
  kubectl delete --cascade=foreground --ignore-not-found --now daemonsets/example-daemonset
```

Author: @sftim I've addressed the comments!

  ```

This simple DaemonSet example introduces key components like init containers and host path volumes,

> which can be expanded upon for more advanced use cases. For more details refer

Reviewer: (nit)
```suggestion
which can be expanded upon for more advanced use cases. For more details refer to
```
